README.md
## Dongju

Download mapbox vector tile specification based resources - styles, tiles, glyphs, and sprites for offline use.
<!--
```
npm install -g dongju
```

## Usage

To download, you need to get a MapBox API Token. You can get one by signing up
for mapbox and requesting a public key from them. It is free. It should start
with `pk`. This is your `token` to be used with mapbox-style-downloader.

To find out the coordinates for the bounding box of the area you'll be downloading
the [boundingbox](https://boundingbox.klokantech.com/) online tool can be used.
Use the `CSV` Copy & Paste option to display the coordinates in the correct format.

To download, use the `download` command like so:

```sh
mapbox-style download mapbox://styles/mapbox/streets-v9 \
  --asar \
  --token='MAPBOX_API_TOKEN' \
  -o styledir \
  -z 8 \
  -Z 13 \
  -b '-60.1364 1.5626 -58.0627 3.475'
```

You need to provide a Mapbox Style URL as the first argument to `download`.
There are many of them, but this is an example: `mapbox://styles/mapbox/streets-v9`. You can see the full list on the [MapBox website](https://docs.mapbox.com/api/maps/styles/).

## Be careful!

This has the capability to send MapBox a lot of requests. Try not to get your
token throttled! Make sure you pick reasonable bounding boxes and zoom levels,
because the data can get very large, very fast.

## Serve styles

To serve the styles, you can use a static http server. MapBox isn't that picky
on how you serve your tiles. There is a built-in example with this command-line
tool. Just give the path to your `styledir` like so:

```
mapbox-style serve styledir
```

And you will be able to see your map, being served offline from the tiles,
sprites, and glyphs you just downloaded.

## API

```
USAGE: mapbox-style <command> [options]

  download STYLE_URL [options]
    -a, --asar          export tile sources as asar archives
    -b, --bounds        bounding box in the form of "lat, lon, lat, lon"
    -o, --output        the output path for the styles
    -z, --minzoom       the minimum zoom for tile downloading [1,16]
    -Z, --maxzoom       the maximum zoom for tile downloading [1,16]
    -t, --token         your MapBox API token
    -u, --minutf        minimum UTF-8 char code to download glyphs for
    -U, --maxutf        maximum UTF-8 char code to download glyphs for
    --style             only download the style.json
    --glyphs            download glyphs
    --tiles             download tiles
    --sprites           download sprites
  
  By default all resources (style.json, glyphs, tiles and sprites) are downloaded.
  Passing any of the --style, --glyphs, --tiles, --sprites options will result
  in only the resources specified being downloaded. This can be helpful if, for
  example, only the style has changes, but the tile data remains the same, so
  there is no need to download it a second time. Note that the style.json is
  _always_ downloaded, since it is needed to get the URLs for other resources.

  serve
    -p, --port          the port to use for the server

  help
    see this help text
```
-->


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.5.2](https://github.com/digidem/mapbox-style-downloader/compare/v2.5.1...v2.5.2) (2021-08-09)


### Bug Fixes

* Update viewer to use mapbox-gl@1.4.1 ([76f5f23](https://github.com/digidem/mapbox-style-downloader/commit/76f5f2394ce0ef4732c9d7cfa3110859a56e0c63))

### [2.5.1](https://github.com/digidem/mapbox-style-downloader/compare/v2.5.0...v2.5.1) (2021-08-09)


### Bug Fixes

* Update mapbox-gl@1.4 to match version in Mapeo Desktop ([5f59931](https://github.com/digidem/mapbox-style-downloader/commit/5f599318fce470c8db75e75772adaf0ff907ab66))

## [2.5.0](https://github.com/digidem/mapbox-style-downloader/compare/v2.4.2...v2.5.0) (2021-07-29)


### Features

* Options to only download specific resources ([9928151](https://github.com/digidem/mapbox-style-downloader/commit/9928151ee6cb480c8f013e7cb9ec9d604857d3e7))


lib/serve.js
const ecstatic = require('ecstatic')
const http = require('http')
const url = require('url')
const getport = require('getport')
const fs = require('fs')
const path = require('path')
const asar = require('asar')

module.exports = serve

// /tiles/:tilesetId/.*
var tilePathRe = /^\/tiles\/((?:[^/]+?))\/(.*)$/i

function serve (root, styleFile, port) {
  var server = http.createServer(function (req, res) {
    var pathname = url.parse(req.url).pathname
    var tilePathMatch = tilePathRe.exec(pathname)
    if (tilePathMatch) {
      console.log('tileserve')
      return serveTile(req, res, root, tilePathMatch)
    }
    if (pathname === '/style.json') {
      return serveStyleFile(req, res, styleFile)
    }
    ecstatic({ root: root, cors: true })(req, res)
  })

  getport(port, function (err, port) {
    if (err) return console.error(err)
    server.listen(port, function () {
      console.log('Listening on port:', port)
    })
  })

  return server
}

function serveStyleFile (req, res, styleFile) {
  fs.stat(styleFile, function (err, stat) {
    if (err) console.error(err)
    fs.readFile(styleFile, 'utf8', function (err, data) {
      if (err) console.error(err)
      data = Buffer.from(data.replace(/\{host\}/gm, 'http://' + req.headers.host))
      res.setHeader('content-type', 'application/json; charset=utf-8')
      res.setHeader('last-modified', (new Date(stat.mtime)).toUTCString())
      res.setHeader('content-length', data.length)
      res.setHeader('Access-Control-Allow-Headers', 'Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since')
      res.setHeader('Access-Control-Allow-Origin', '*')
      res.statusCode = 200
      res.write(data)
      res.end()
    })
  })
}

function serveTile (req, res, root, match) {
  var pathname = url.parse(req.url).pathname
  if (pathname.endsWith('.pbf') || pathname.endsWith('.mvt')) {
    // Vector tiles are gzip at rest, so need to be served as gzip
    // Raster tiles should not be served as gzip
    res.setHeader('content-encoding', 'gzip')
  }
  var tilesetId = match[1]
  var asarPath = path.join(root, 'tiles', tilesetId + '.asar')
  fs.access(asarPath, fs.constants.R_OK, function (err) {
    if (err) console.log('no asar')
    // if no asar file, serve tile as static file
    if (err) return ecstatic({ root: root, cors: true })(req, res)
    console.log('asar', asarPath, match[2])
    var tilePath = match[2]
    serveAsarTile(req, res, asarPath, tilePath)
  })
}

function serveAsarTile (req, res, asarPath, tilePath) {
  try {
    var buf = asar.extractFile(asarPath, tilePath)
    var ext = path.parse(tilePath).ext
    console.log(buf)
    var mime
    switch (ext) {
      case 'png': mime = 'image/png'; break
      case 'jpg': mime = 'image/jpg'; break
    }
    if (mime) res.setHeader('content-type', mime)
    res.setHeader('Access-Control-Allow-Origin', '*')
    res.end(buf)
  } catch (e) {
    res.statusCode = 404
    res.end()
  }
}


lib/glyphs.js
const request = require('request')
const traverse = require('traverse')
const queue = require('d3-queue').queue
const fs = require('fs')
const path = require('path')
const mkdirp = require('mkdirp')
const pump = require('pump')

const normalizeURL = require('./mapbox').normalizeGlyphsURL

const DEFAULT_FONTSTACK = ["Open Sans Regular", "Arial Unicode MS Regular"]

module.exports = function (style, outDir, accessToken, range, cb) {
  if (!style.glyphs) return cb(new Error('no glyph url defined in style'))
  console.log('Downloading glyphs for UTF char range ' + range[0] + '-' + range[1] + '...')
  mkdirp(outDir)
  var fontStacks = getFontStacks(style)
  downloadGlyphs(style.glyphs, fontStacks, outDir, accessToken, range, cb)
}

function getFontStacks (style) {
  var fontStacks = {}
  fontStacks[DEFAULT_FONTSTACK.join(',')] = true
  traverse(style).forEach(function (x) {
    if (this.key === 'text-font') {
      if (Array.isArray(x)) {
        if (x[0] === 'step' && x[2] && x[2][0] === 'literal') {
          if (Array.isArray(x[2][1])) {
            fontStacks[x[2][1].join(',')] = true
          } else {
            fontStacks[x[2][1]] = true
          }
        } else if (x[0] === 'literal') {
          if (Array.isArray(x[1])) {
            fontStacks[x[1].join(',')] = true
          } else {
            fontStacks[x[1]] = true
          }
        } else {
          fontStacks[x.join(',')] = true
        }
      } else if (typeof x === 'string') {
        fontStacks[x] = true
      } else if (x.stops) {
        x.stops.forEach(stop => {
          var stack = Array.isArray(stop[1]) ? stop[1].join(',') : stop[1]
          fontStacks[stack] = true
        })
      }
    }
  })
  return Object.keys(fontStacks)
}

function downloadGlyphs (urlTemplate, fontStacks, outDir, accessToken, range, cb) {
  urlTemplate = normalizeURL(urlTemplate, accessToken)
  var q = queue(10)
  fontStacks.forEach(stack => {
    var stackDir = path.join(outDir, stack)
    mkdirp.sync(stackDir)
    var rangeStr
    var url
    var outFilepath
    for (var i = 0; i < 65536; (i = i + 256)) {
      if (range[1] < i || range[0] > i + 256) continue
      rangeStr = i + '-' + (i + 255)
      url = glyphUrl(stack, rangeStr, urlTemplate)
      outFilepath = path.join(stackDir, rangeStr + '.pbf')
      q.defer(downloadToFile, url, outFilepath)
    }
  })
  q.awaitAll(cb)
}

function downloadToFile (url, filepath, cb) {
  console.log(url)
  var req = request({url: url, gzip: true})
  pump(req, fs.createWriteStream(filepath), cb)
}

/**
 * Use CNAME sharding to load a specific glyph range over a randomized
 * but consistent subdomain.
 * @param {string} fontstack comma-joined fonts
 * @param {string} range comma-joined range
 * @param {url} url templated url
 * @param {string} [subdomains=abc] subdomains as a string where each letter is one.
 * @returns {string} a url to load that section of glyphs
 * @private
 */
function glyphUrl (fontstack, range, url, subdomains) {
  subdomains = subdomains || 'abc'

  return url
        .replace('{s}', subdomains[fontstack.length % subdomains.length])
        .replace('{fontstack}', fontstack)
        .replace('{range}', range)
}


lib/mapbox.js
'use strict'
// from https://github.com/mapbox/mapbox-gl-js/blob/495a695cb63ed55b4b464da83bfd085738ee57ef/src/util/mapbox.js

const config = {
  API_URL: 'https://api.mapbox.com',
  REQUIRE_ACCESS_TOKEN: true,
  ACCESS_TOKEN: null
}

const help = 'See https://www.mapbox.com/api-documentation/#access-tokens'

function makeAPIURL (urlObject, accessToken) {
  const apiUrlObject = parseUrl(config.API_URL)
  urlObject.protocol = apiUrlObject.protocol
  urlObject.authority = apiUrlObject.authority

  if (!config.REQUIRE_ACCESS_TOKEN) return formatUrl(urlObject)

  accessToken = accessToken || config.ACCESS_TOKEN
  if (!accessToken) { throw new Error(`An API access token is required to use Mapbox GL. ${help}`) }
  if (accessToken[0] === 's') { throw new Error(`Use a public access token (pk.*) with Mapbox GL, not a secret access token (sk.*). ${help}`) }

  urlObject.params.push(`access_token=${accessToken}`)
  return formatUrl(urlObject)
}

function makeACKEYURL(url, accessToken) {
  return url.replace(/YOUR\-API\-KEY/, accessToken)
}

function isMapboxURL (url) {
  return url.indexOf('mapbox:') === 0
}

exports.isMapboxURL = isMapboxURL

exports.normalizeStyleURL = function (url, accessToken) {
  if (!isMapboxURL(url)) return makeACKEYURL(url, accessToken)
  const urlObject = parseUrl(url)
  urlObject.path = `/styles/v1${urlObject.path}`
  return makeAPIURL(urlObject, accessToken)
}

exports.normalizeGlyphsURL = function (url, accessToken) {
  if (!isMapboxURL(url)) return makeACKEYURL(url, accessToken)
  const urlObject = parseUrl(url)
  urlObject.path = `/fonts/v1${urlObject.path}`
  return makeAPIURL(urlObject, accessToken)
}

exports.normalizeSourceURL = function (url, accessToken) {
  if (!isMapboxURL(url)) return makeACKEYURL(url, accessToken)
  const urlObject = parseUrl(url)
  urlObject.path = `/v4/${urlObject.authority}.json`
    // TileJSON requests need a secure flag appended to their URLs so
    // that the server knows to send SSL-ified resource references.
  urlObject.params.push('secure')
  return makeAPIURL(urlObject, accessToken)
}

exports.normalizeSpriteURL = function (url, format, extension, accessToken) {
  const urlObject = parseUrl(url)
  if (!isMapboxURL(url)) {
    urlObject.path += `${format}${extension}`
    return makeACKEYURL(formatUrl(urlObject), accessToken)
  }
  urlObject.path = `/styles/v1${urlObject.path}/sprite${format}${extension}`
  return makeAPIURL(urlObject, accessToken)
}

const imageExtensionRe = /(\.(png|jpg)\d*)(?=$)/

exports.normalizeTileURL = function (tileURL, sourceURL, tileSize, opts) {
  opts = opts || {}
  if (!sourceURL || !isMapboxURL(sourceURL)) return tileURL

  const urlObject = parseUrl(tileURL)

    // The v4 mapbox tile API supports 512x512 image tiles only when @2x
    // is appended to the tile URL. If `tileSize: 512` is specified for
    // a Mapbox raster source force the @2x suffix even if a non hidpi device.
  const suffix = opts.devicePixelRatio >= 2 || tileSize === 512 ? '@2x' : ''
  const extension = opts.supportsWebp ? '.webp' : '$1'
  urlObject.path = urlObject.path.replace(imageExtensionRe, `${suffix}${extension}`)

  replaceTempAccessToken(urlObject.params)
  return formatUrl(urlObject)
}

function replaceTempAccessToken (params) {
  for (let i = 0; i < params.length; i++) {
    if (params[i].indexOf('access_token=tk.') === 0) {
      params[i] = `access_token=${config.ACCESS_TOKEN || ''}`
    }
  }
}

const urlRe = /^(\w+):\/\/([^/?]*)(\/[^?]+)?\??(.+)?/

function parseUrl (url) {
  const parts = url.match(urlRe)
  if (!parts) {
    throw new Error('Unable to parse URL object')
  }
  return {
    protocol: parts[1],
    authority: parts[2],
    path: parts[3] || '/',
    params: parts[4] ? parts[4].split('&') : []
  }
}

function formatUrl (obj) {
  const params = obj.params.length ? `?${obj.params.join('&')}` : ''
  return `${obj.protocol}://${obj.authority}${obj.path}${params}`
}


lib/download.js
const debug = require('debug')('mapbox-style-downloader')
const request = require('request')
const mkdirp = require('mkdirp')
const path = require('path')
const queue = require('d3-queue').queue
const assign = require('object-assign')
const fs = require('fs-extra')

const sanitize = require('./sanitize-id')
const normalizeStyleURL = require('./mapbox').normalizeStyleURL
const downloadGlyphs = require('./glyphs')
const downloadSprites = require('./sprites')
const downloadTiles = require('./tiles')

const defaults = {
  glyphs: false,
  sprites: false,
  tiles: false,
  style: false,
  minzoom: 1,
  maxzoom: 14,
  bounds: '-180.0 -85.06 180.0 85.06',
  sources: null,
  range: [0, 65535],
  asar: false
}

module.exports = download

function download (styleUrl, accessToken, outDir, opts, cb) {
  opts = Object.assign({}, defaults, opts)
  opts.bounds = opts.bounds.split(/,|\s/)

  const all = !opts.glyphs && !opts.sprites && !opts.tiles && !opts.style

  if (typeof opts.minutf === 'number') opts.range[0] = opts.minutf
  if (typeof opts.maxutf === 'number') opts.range[1] = opts.maxutf

  debug('Requesting stylesheet...')
  request(normalizeStyleURL(styleUrl, accessToken), { json: true }, function (err, res, stylesheet) {
    if (err) return cb(err)
    if (res.body.message) return cb(new Error(res.body.message))
    var offlineStylesheet = assign({}, stylesheet)
    var q = queue(1)

    mkdirp.sync(outDir)
    if (opts.glyphs || all) {
      q.defer(downloadGlyphs, stylesheet, path.join(outDir, 'fonts'), accessToken, opts.range)
      offlineStylesheet.glyphs = '{host}/fonts/{fontstack}/{range}.pbf'
    }
    if (opts.sprites || all) {
      q.defer(downloadSprites, stylesheet.sprite, path.join(outDir, 'sprites'), accessToken)
      offlineStylesheet.sprite = '{host}/sprites/sprite'
    }
    if (opts.tiles || all) {
      q.defer(downloadSources)
    }
    q.awaitAll(function (err) {
      if (err) return cb(err)
      offlineStylesheet.zoom = opts.minzoom
      const centerLon = opts.bounds[0] + (opts.bounds[2] - opts.bounds[0]) / 2
      const centerLat = opts.bounds[1] + (opts.bounds[3] - opts.bounds[1]) / 2
      offlineStylesheet.center = [centerLon, centerLat]
      fs.writeFileSync(path.join(outDir, 'style.json'), JSON.stringify(offlineStylesheet, null, 2))
      fs.copySync(path.join(__dirname, '../index.html'), path.join(outDir, 'index.html'))
      console.log('Style resources successfully downloaded to ', outDir)
    })

    function downloadSources (cb) {
      console.log('Organizing sources...')
      var sourceQ = queue(1)
      var sourceKeys = Object.keys(stylesheet.sources)

      sourceKeys
        .forEach(function (key, i) {
          var sourceUrl = stylesheet.sources[key].url
          sourceQ.defer(downloadTiles, sourceUrl, path.join(outDir, 'tiles'), accessToken, opts)
        })
      sourceQ.awaitAll(function (err, tilejsons) {
        if (err) return cb(err)
        offlineStylesheet.sources = sourceKeys.reduce(function (acc, key, i) {
          acc[key] = tilejsons[i]
          var sanitizedId = sanitize(acc[key].id)
          acc[key].tiles = acc[key].tiles.map(function (url) { return '{host}/tiles/' + sanitizedId + url })
          acc[key].type = stylesheet.sources[key].type
          return acc
        }, {})
        cb()
      })
    }
  })
}


lib/tiles.js
process.env.UV_THREADPOOL_SIZE = process.env.UV_THREADPOOL_SIZE || Math.ceil(Math.max(4, require('os').cpus().length * 1.5))

const debug = require('debug')('mapbox-style-downloader')
const mkdirp = require('mkdirp')
const path = require('path')
const parseUrl = require('url').parse
const tilelive = require('tilelive-streaming')(require('@mapbox/tilelive'))
const asar = require('asar')
const rimraf = require('rimraf')

const tileJsonClass = require('@mapbox/tilejson')
tileJsonClass.registerProtocols(tilelive)
require('tilelive-file').registerProtocols(tilelive)
const request = require('requestretry');
const Agent = require('agentkeepalive');
const agent = new Agent({
  maxSockets: 128,
  keepAliveTimeout: 30000
});
const httpsagent = new Agent.HttpsAgent({
  maxSockets: 128,
  keepAliveTimeout: 30000
});
tileJsonClass.prototype.get = function(url, callback) {
  request({
    url: url,
    encoding: null,
    timeout: this.timeout,
    headers: {Connection: 'Keep-Alive', Origin: 'http://localhost:8000'}, // For adding Origin
    agent: url.indexOf('https:') === 0 ? httpsagent : agent,
    maxAttempts: 2,
    retryDelay: 0
  }, function (err, res, buffer) {
    if (!err && res.statusCode !== 200) {
      err = new Error('Server returned HTTP ' + res.statusCode);
      err.statusCode = res.statusCode;
    }
    callback(err, buffer, res && res.headers);
  });
};

const sanitize = require('./sanitize-id')
const normalizeSourceURL = require('./mapbox').normalizeSourceURL

module.exports = function (url, outDir, accessToken, opts, cb) {
  opts = opts || {}
  opts.minzoom = 'minzoom' in opts ? opts.minzoom : 0
  opts.maxzoom = 'maxzoom' in opts ? opts.maxzoom : 16
  if (!opts.bounds) console.warn('warning: no bounds set, downloading entire world')
  opts.bounds = opts.bounds || [-180, -85.0511, 180, 85.0511]

  debug('waiting for tilelive.load')
  tilelive.load('tilejson+' + normalizeSourceURL(url, accessToken), function (err, source) {
    if (err) return cb(err)
    var tilejson = source.data
    var sourceId = tilejson.id
    outDir = path.join(outDir, sanitize(sourceId))
    mkdirp.sync(outDir)

    opts.minzoom = Math.max(opts.minzoom, tilejson.minzoom || 0)
    opts.maxzoom = Math.min(opts.maxzoom, tilejson.maxzoom || 20)
    if (opts.minzoom > opts.maxzoom) {
      if (opts.minzoom === tilejson.minzoom) opts.maxzoom = opts.minzoom
      else opts.minzoom = opts.maxzoom
    }
    tilejson.minzoom = opts.minzoom
    tilejson.maxzoom = opts.maxzoom

    var tilePath = parseUrl(tilejson.tiles[0] || tilejson.tiles).pathname
    var filetype = path.parse(tilePath).base.replace('%7By%7D', '').replace(/^\./, '')
    return tilelive.load('file://' + outDir, function (err, sink) {
      if (err) return cb(err)
      sink.filetype = filetype
      source.createReadStream(opts)
        .pipe(sink.createWriteStream())
        .on('tile', function (tile) {
          if (!opts.quiet) {
            console.log('%d/%d/%d\t%d', tile.z, tile.x, tile.y, tile.length)
          }
        })
        .on('error', cb)
        .on('end', function () {
          tilejson.tiles = ['/{z}/{x}/{y}.' + filetype]
          if (!opts.asar) return cb(null, tilejson)
          asar.createPackage(outDir, outDir + '.asar', function (err) {
            if (err) return cb(err)
            rimraf(outDir, function (err) {
              cb(err, tilejson)
            })
          })
        })
    })
  })
}
